apis/v1alpha2/config.go

```diff
@@ -0,0 +1,147 @@
+func (in *AnyConfig) DeepCopyInto(out *AnyConfig) {
+	*out = *in
+	if in.Object != nil {
+		in, out := &in.Object, &out.Object
+		*out = make(map[string]interface{}, len(*in))
+		for key, val := range *in {
+			(*out)[key] = val
+		}
+	}
+}
+
+// DeepCopy is an autogenerated deepcopy function, copying the receiver, creating a new AnyConfig.
+func (in *AnyConfig) DeepCopy() *AnyConfig {
+	if in == nil {
+		return nil
+	}
+	out := new(AnyConfig)
+	in.DeepCopyInto(out)
+	return out
+}
+
+var _ json.Marshaler = &AnyConfig{}
+var _ json.Unmarshaler = &AnyConfig{}
+
+// UnmarshalJSON implements an alternative parser for this field.
+func (c *AnyConfig) UnmarshalJSON(b []byte) error {
+	vals := map[string]interface{}{}
+	if err := json.Unmarshal(b, &vals); err != nil {
+		return err
+	}
+	c.Object = vals
+	return nil
+}
+
+// MarshalJSON specifies how to convert this object into JSON.
+func (c *AnyConfig) MarshalJSON() ([]byte, error) {
+	if c == nil {
+		return []byte("{}"), nil
+	}
+	return json.Marshal(c.Object)
+}
+
+// Config encapsulates collector config.
+type Config struct {
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Receivers AnyConfig `json:"receivers"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Exporters AnyConfig `json:"exporters"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Processors *AnyConfig `json:"processors,omitempty"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Connectors *AnyConfig `json:"connectors,omitempty"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Extensions *AnyConfig `json:"extensions,omitempty"`
+	Service    Service    `json:"service"`
+}
+
+type Service struct {
+	Extensions *[]string `json:"extensions,omitempty"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Telemetry *AnyConfig `json:"telemetry,omitempty"`
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Pipelines AnyConfig `json:"pipelines"`
+}
+
+// Returns null objects in the config.
+func (c Config) nullObjects() []string {
+	var nullKeys []string
+	if nulls := hasNullValue(c.Receivers.Object); len(nulls) > 0 {
+		nullKeys = append(nullKeys, addPrefix("receivers.", nulls)...)
+	}
+	if nulls := hasNullValue(c.Exporters.Object); len(nulls) > 0 {
+		nullKeys = append(nullKeys, addPrefix("exporters.", nulls)...)
+	}
+	if c.Processors != nil {
+		if nulls := hasNullValue(c.Processors.Object); len(nulls) > 0 {
+			nullKeys = append(nullKeys, addPrefix("processors.", nulls)...)
+		}
+	}
+	if c.Extensions != nil {
+		if nulls := hasNullValue(c.Extensions.Object); len(nulls) > 0 {
+			nullKeys = append(nullKeys, addPrefix("extensions.", nulls)...)
+		}
+	}
+	if c.Connectors != nil {
+		if nulls := hasNullValue(c.Connectors.Object); len(nulls) > 0 {
+			nullKeys = append(nullKeys, addPrefix("connectors.", nulls)...)
+		}
+	}
+	// Make the return deterministic. The config uses maps therefore processing order is non-deterministic.
+	sort.Strings(nullKeys)
+	return nullKeys
+}
+
+func hasNullValue(cfg map[string]interface{}) []string {
```

Comment:
this is great – in the future in the defaulting webhook we could set the empty map using this as well :)